nntrainer/tensor/tensor.cpp

```diff
@@ -239,6 +239,62 @@ void Tensor::setRandUniform(float min, float max) {
     std::uniform_real_distribution<float>(min, max));
 }
 
+void Tensor::initialize() {
+  if (empty() || !isAllocated())
```

Comment:
early return here if initializer is none as well?

Reply:
and if empty(), !isAllocated, shouldn't it be throwing?

Reply:
> early return here if initializer is none as well?

will add

Reply:
> and if empty(), !isAllocated, shouldn't it be throwing?

- in !isAllocated(), it will note the initialize value, and whenever access, the tensor will be initialized (allocate will initialize). This allows delayed initialized till the memory is allocated.
- if empty(), the above behavior will be followed, but here, we can throw as the tensor shape info is not available.

Reply:
call of initialize() is missing in allocate() (which I thought I did add). I will add it.

Reply:
So, is initialize must come after allocate? or vice versa? or have no order?

`empty()` previously was checking on `unintialized()` checks on the `size()`, which does not require `allocate()`

if allocate -> initialize
   allocate should not initialize but should be always separate call for the consistent semantics or initialize should be a private function? But it seems, allocate() need tensor dimension fixed so initialize has to be done first?

We might need different naming or firmer semantics. Delaying also sounds good though.

Reply:
initialize() ideally must come after allocate(), but calling before it is a no-op.
initialize(init=...) has an extra effect of setting which initializer to use.

When allocate() will be called, the memory will be allocated and tensor will be initialize() at that time (by calling intiailize() internally).

Reply:
> which follows different question, is uninitialized() == empty() ?

`uninitialized()` was confusing because of its name. So, it is removed.
`empty()` means size is 0.
`isAllocated()` means memory is allocated.

Reply:
Looks like there is actually three things to be done.

1. setting tensor dimension (it's not initialize nor allocate)
2. initialize()
3. allocate()

If we will to delay the initialize(), this has to be done with a state and we shouldn't always initialize(). and could you check if  `empty()` have to throw at `allocate()`?

```cpp
void initialize() {
  if (unallocated() ) { 
    this->book_initialize = true; /// or throw
  }
}

void allocate() {
  if(this->book_initialize) {
    initialize();
  }
}
```

Other option might be fixating the order?

Reply:
The delay of initialization is done with the allocation. Once allocation is done, initialization will always be done.
In default case, the initialization is NONE, which results in no-op.

Reply:
I mean even if there is a initializer is not NONE, it shouldn't be done on allocation, it has to be done only when requested.

Reply:
In case of lazy allocation, requesting the initialization again later is a burden for the caller.

>  it shouldn't be done on allocation, it has to be done only when requested.

can you expand why it shouldn't be done on allocation and only when requested?

Reply:
```cpp
TEST(nntrainer_Tensor, initialize_03_p) {
  nntrainer::Tensor t({1, 2, 3, 4}, false,
                      nntrainer::Tensor::Initializer::ONES);
  t.allocate();

  nntrainer::Tensor golden(1, 2, 3, 4);
  golden.setValue(1);

  EXPECT_EQ(golden, t);
}
```

above didn't call `initialize()` but allocated.

case 1. if initialize() only takes places when allocating, we don't need to call `initialize()` but can be private.
case 2. if initialize() takes multiple times (say, every iteration), some line of code will contain `t.initialize()`, so we will be calling initialize() twice for the first iteration.

Moreover, if we don't initialize but allocate implicitly call initialize(I mean not delayed initialize() but we don't call initialize) this will have same effect as #258 points out

Reply:
> above didn't call initialize() but allocated + initialized.

The code requested a tensor which must be initialized with ONES when accessed. So, the above need not call initialize explicitly.

The initialize is called every time with allocating, and only once by design. This ensures that once the tensor is allocated, it is initialized if needed (refer to https://github.com/nnstreamer/nntrainer/pull/1031#discussion_r597551354 for more details).

> case 1. if initialize() only takes place when allocating, we don't need to call initialize() so it can be private.

initialize takes place when allocating doesn't mean it can't be done again. `initialize()` is public for users to re-initialize their tensors whenever they want.

> case 2. if initialize() takes multiple times (say, every iteration), some line of code will contain t.initialize(), so we will be calling initialize() twice for the first iteration.

This is not by design for tensors but can be done by the caller. If a tensor is to be initialized every iteration and the design is leading to double initialization, use the below sample code:
```
Tensor t = Tensor({SHAPE}, false, Tensor::Initialize::NONE);
.....
t.allocate();
.....
for (iter : iterations)
  t.inialize(Tensor::Iinitializer::ONES);
```

> Moreover, if we don't initialize but allocate implicitly call initialize(I mean not delayed initialize() but we don't call initialize) this will have same effect as #258 points out

#258 had an overhead of zeroing out every tensor. But the current default initialization is NONE. So, allocate implicitly call initialize will have no overhead. I dont understand how these two are related.

Reply:
Seeing https://github.com/nnstreamer/nntrainer/pull/1031#discussion_r597551354 I got the intention :) So it's basically RAII to some extent.

I'd like to suggest hiding `initialize()` but leaving `initialize(Tensor::Initializer initializer)` to only allow below cases.

```cpp
Tensor t = Tensor({SHAPE}, false, Tensor::Initialize::NONE);
.....
t.allocate();
.....
for (iter : iterations)
  t.inialize(Tensor::Iinitializer::ONES);
```

Reply:
> I'd like to suggest hiding initialize() 

This is open for the purpose of gradient, to be able to zero it without having to explicitly use its initializer or storing it somewhere.

Reply:
Cool